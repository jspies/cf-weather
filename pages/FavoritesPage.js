import Layout from '../components/Layout';
import Favorites from '../components/Favorites';
import React from 'react';
import { debounce } from 'lodash';
import {
  Dropdown,
  DropdownLink,
  DropdownSeparator
} from '@cloudflare/component-dropdown';
import { LabeledInput } from '@cloudflare/component-input';
import { createComponent } from '@cloudflare/style-container';
import PropTypes from 'prop-types';

const Selector = createComponent(() => ({
  whieSpace: 'nowrap',
  display: 'inline-block',
  verticalAlign: 'middle',
  position: 'relative'
}));

const FavoritesPage = class Index extends React.Component {
  /**
   * favorites [array] - the cities in the favorite list
   * cities [array] - the cities in the search dropdown
   * dropdownOpen [boolean] - whether or not the search dropdown is open
   */
  constructor(props) {
    super(props);

    this.state = {
      favorites: [],
      cities: [],
      dropdownOpen: false,
    };

    this.debounceSearchCities = debounce(this.searchCities, 500);
  }

  async componentDidMount() {
    const { getFavorites } = this.props;
    const favorites = await getFavorites();
    this.setState({
      favorites
    });
  }

  /**
   * searches cities by the search criteria and updates the state accordingly
   * @param city - the search criteria
   */
  async searchCities(city) {
    const { getCitiesMatchingQuery } = this.props;
    let cities = await getCitiesMatchingQuery(city);
    cities.sort((a, b) => {
      if (a.name < b.name)
        return -1;
      if (a.name > b.name)
        return 1;
      return 0;
    });
    if (cities.length > 20) {
      cities.length = 20;
    }
    this.setState({
      cities,
      dropdownOpen: cities.length > 0,
    });
  }

  /**
   * adds the selected city to the favorites page
   * @param city - the selected city
   */
  async addToFavorites(city) {
    const { addFavorite, getFavorites } = this.props;
    await addFavorite(city);
    const favorites = await getFavorites();
    this.setState({
      favorites,
      dropdownOpen: false,
    });
  }

  render() {
    return (
      <Layout>
        <Selector>
          <LabeledInput
            label="Search for a city"
            mb={0}
            name="citySearch"
            id="citySearch"
            onChange={(e) => {
              this.debounceSearchCities(e.target.value);
            }}
          />
          {this.state.dropdownOpen &&
            <Dropdown
              closed
              onClose={() => {
                this.setState({
                  dropdownOpen: false,
                });
              }}
            >
            {this.state.cities.map((city) => {
              return (
                <DropdownLink
                  key={city.id}
                  className="dropdown-link"
                  onClick={() => this.addToFavorites(city.name)}
                >
                  {city.name}
                </DropdownLink>
              );
            })}
            </Dropdown>
          }
        </Selector>
        <Favorites favorites={this.state.favorites} />
      </Layout>
    );
  }
};

FavoritesPage.propTypes = {
  getFavorites: PropTypes.func.isRequired,
  addFavorite: PropTypes.func.isRequired,
  getCitiesMatchingQuery: PropTypes.func.isRequired
};
export default FavoritesPage;
