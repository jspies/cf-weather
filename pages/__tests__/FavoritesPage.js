import { felaShallow } from '@cloudflare/util-testing';
import FavoritesPage from '../FavoritesPage';

const addFavorite = jest.fn().mockImplementation(() => {
  return Promise.resolve({ id: 1, city: 'austin' });
});

const getFavorites = jest.fn().mockImplementation(() => {
  return Promise.resolve([
    { id: 0, city: 'london' },
    { id: 1, city: 'austin' },
  ]);
});

const getCitiesMatchingQuery = jest.fn().mockImplementation(() => {
  return Promise.resolve([
    {
      id: 4254010,
      name: "Austin",
      country: "US",
      coord: {
        lon: -85.808029,
        lat: 38.758389
      }
    },
    {
      id: 1001775,
      name: "Glen Austin",
      country: "ZA",
      coord: {
        lon: 28.15,
        lat: -25.966669
      }
    },
  ]);
});

const defaultProps = {
  addFavorite,
  getFavorites,
  getCitiesMatchingQuery
};

describe('FavoritesPage', () => {
  it('should render the favorites', done => {
    const { snapshot, wrapper } = felaShallow(
      <FavoritesPage {...defaultProps} />
    );
    expect(snapshot(wrapper)).toMatchSnapshot();

    // need to check snapshot after the getFavorites call's promise has resolved
    setTimeout(() => {
      wrapper.update();
      expect(snapshot(wrapper)).toMatchSnapshot();
      expect(defaultProps.getFavorites).toHaveBeenCalled();
      expect(wrapper.state().favorites).toEqual([
        { id: 0, city: 'london' },
        { id: 1, city: 'austin' },
      ]);
      done();
    });
  });

  it('should call debounceSearchCities', done => {
    const { snapshot, wrapper } = felaShallow(
      <FavoritesPage {...defaultProps} />
    );

    const debounceSpy = jest.spyOn(wrapper.instance(), 'debounceSearchCities');
    wrapper.find('#citySearch').simulate('change', { target: { value: 'austin' } });
    setTimeout(() => {
      wrapper.update();
      expect(debounceSpy).toHaveBeenCalled();
      done();
    });
  });

  it('should search and add cities to the search dropdown', done => {
    const { snapshot, wrapper } = felaShallow(
      <FavoritesPage {...defaultProps} />
    );

    wrapper.instance().searchCities('austin');
    setTimeout(() => {
      wrapper.update();
      expect(snapshot(wrapper)).toMatchSnapshot();
      expect(defaultProps.getCitiesMatchingQuery).toHaveBeenCalledWith('austin');
      expect(wrapper.state().dropdownOpen).toEqual(true);
      expect(wrapper.state().cities).toEqual([
        {
          id: 4254010,
          name: "Austin",
          country: "US",
          coord: {
            lon: -85.808029,
            lat: 38.758389
          }
        },
        {
          id: 1001775,
          name: "Glen Austin",
          country: "ZA",
          coord: {
            lon: 28.15,
            lat: -25.966669
          }
        },
      ]);
      done();
    });
  });

  it('should add city to favorites on click of a city in the dropdown', done => {
    const { snapshot, wrapper } = felaShallow(
      <FavoritesPage {...defaultProps} />
    );

    const addToFavoritesSpy = jest.spyOn(FavoritesPage.prototype, 'addToFavorites');
    wrapper.setState({
      dropdownOpen: true,
      cities: [
        {
          id: 4254010,
          name: "Austin",
          country: "US",
          coord: {
            lon: -85.808029,
            lat: 38.758389
          }
        },
        {
          id: 1001775,
          name: "Glen Austin",
          country: "ZA",
          coord: {
            lon: 28.15,
            lat: -25.966669
          }
        },
      ],
    }, () => {
      wrapper.update();
      expect(snapshot(wrapper)).toMatchSnapshot();
      wrapper.find('.dropdown-link').first().simulate('click');
      setTimeout(() => {
        expect(addToFavoritesSpy).toHaveBeenCalledWith('Austin');
        expect(defaultProps.addFavorite).toHaveBeenCalledWith('Austin');
        expect(wrapper.state().dropdownOpen).toEqual(false);
        expect(wrapper.state().favorites).toEqual([
          { id: 0, city: 'london' },
          { id: 1, city: 'austin' },
        ]);
        done();
      });
    });
  });
});
