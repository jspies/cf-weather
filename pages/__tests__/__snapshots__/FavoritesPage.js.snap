// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesPage should add city to favorites on click of a city in the dropdown 1`] = `
Object {
  "component": <Layout>
    <FelaComponent(div)>
      <LabeledInput
        id="citySearch"
        label="Search for a city"
        mb={0}
        name="citySearch"
        onChange={[Function]}
      />
      <FelaComponent(Dropdown)
        align="left"
        closed={true}
        onClose={[Function]}
      >
        <FelaComponent(DropdownLink)
          className="dropdown-link"
          key="4254010"
          onClick={[Function]}
        >
          Austin
        </FelaComponent(DropdownLink)>
        <FelaComponent(DropdownLink)
          className="dropdown-link"
          key="1001775"
          onClick={[Function]}
        >
          Glen Austin
        </FelaComponent(DropdownLink)>
      </FelaComponent(Dropdown)>
    </FelaComponent(div)>
    <Favorites
      favorites={Array []}
    />
  </Layout>,
  "styles": "

",
}
`;

exports[`FavoritesPage should render the favorites 1`] = `
Object {
  "component": <Layout>
    <FelaComponent(div)>
      <LabeledInput
        id="citySearch"
        label="Search for a city"
        mb={0}
        name="citySearch"
        onChange={[Function]}
      />
    </FelaComponent(div)>
    <Favorites
      favorites={Array []}
    />
  </Layout>,
  "styles": "

",
}
`;

exports[`FavoritesPage should render the favorites 2`] = `
Object {
  "component": <Layout>
    <FelaComponent(div)>
      <LabeledInput
        id="citySearch"
        label="Search for a city"
        mb={0}
        name="citySearch"
        onChange={[Function]}
      />
    </FelaComponent(div)>
    <Favorites
      favorites={
        Array [
          Object {
            "city": "london",
            "id": 0,
          },
          Object {
            "city": "austin",
            "id": 1,
          },
        ]
      }
    />
  </Layout>,
  "styles": "

",
}
`;

exports[`FavoritesPage should search and add cities to the search dropdown 1`] = `
Object {
  "component": <Layout>
    <FelaComponent(div)>
      <LabeledInput
        id="citySearch"
        label="Search for a city"
        mb={0}
        name="citySearch"
        onChange={[Function]}
      />
      <FelaComponent(Dropdown)
        align="left"
        closed={true}
        onClose={[Function]}
      >
        <FelaComponent(DropdownLink)
          className="dropdown-link"
          key="4254010"
          onClick={[Function]}
        >
          Austin
        </FelaComponent(DropdownLink)>
        <FelaComponent(DropdownLink)
          className="dropdown-link"
          key="1001775"
          onClick={[Function]}
        >
          Glen Austin
        </FelaComponent(DropdownLink)>
      </FelaComponent(Dropdown)>
    </FelaComponent(div)>
    <Favorites
      favorites={
        Array [
          Object {
            "city": "london",
            "id": 0,
          },
          Object {
            "city": "austin",
            "id": 1,
          },
        ]
      }
    />
  </Layout>,
  "styles": "

",
}
`;
